RevokeCgn/handler.ts

```diff
@@ -0,0 +1,158 @@
+) => Promise<ReturnTypes>;
+
+const checkExistingCgnStatus = (cgnStatus: CgnStatus) =>
+  CgnRevokedStatus.is(cgnStatus) || CgnCanceledStatus.is(cgnStatus)
+    ? fromLeft<IResponseErrorConflict, CgnStatus>(
+        ResponseErrorConflict(
+          "Cannot revoke the user's cgn because it is already revoked or canceled"
+        )
+      )
+    : taskEither.of<IResponseErrorConflict, CgnStatus>(cgnStatus);
+
+export function RevokeCgnHandler(
+  userCgnModel: UserCgnModel,
+  logPrefix: string = "RevokeCgnHandler"
+): IRevokeCgnHandler {
+  return async (context, fiscalCode, revokationReq) => {
+    const client = df.getClient(context);
+    const revokedCgnStatus: CgnRevokedStatus = {
+      motivation: revokationReq.motivation,
+      revokation_date: new Date(),
+      status: StatusEnum.REVOKED
+    };
+    const orchestratorId = makeUpdateCgnOrchestratorId(
+      fiscalCode,
+      StatusEnum.REVOKED
+    ) as NonEmptyString;
+    return userCgnModel
+      .findLastVersionByModelId([fiscalCode])
+      .mapLeft<IResponseErrorInternal | IResponseErrorNotFound>(() =>
+        ResponseErrorInternal("Cannot retrieve CGN infos for this user")
+      )
+      .chain(maybeUserCgn =>
+        fromEither(
+          fromOption(
+            ResponseErrorNotFound("Not Found", "User's CGN status not found")
+          )(maybeUserCgn)
+        )
+      )
+      .foldTaskEither<ErrorTypes, CgnStatus>(fromLeft, userCgn =>
+        checkExistingCgnStatus(userCgn.status)
+      )
+      .chain(() =>
+        checkUpdateCgnIsRunning(
+          client,
+          fiscalCode,
+          revokedCgnStatus
+        ).foldTaskEither<
+          ErrorTypes,
+          | IResponseSuccessAccepted
+          | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
+        >(
+          response =>
+            response.kind === "IResponseSuccessAccepted"
+              ? taskEither.of(response)
+              : fromLeft(response),
+          () =>
+            tryCatch(
+              () =>
+                client.startNew(
+                  "UpdateCgnOrchestrator",
+                  orchestratorId,
+                  OrchestratorInput.encode({
+                    fiscalCode,
+                    newStatus: revokedCgnStatus
+                  })
+                ),
+              toError
+            ).bimap(
+              err => {
+                context.log.error(
+                  `${logPrefix}|Cannot start UpdateCgnOrchestrator|ERROR=${err.message}`
+                );
+                return ResponseErrorInternal(
+                  "Cannot start UpdateCgnOrchestrator"
+                );
+              },
+              () => {
+                const instanceId: InstanceId = {
+                  id: orchestratorId
+                };
+                return ResponseSuccessRedirectToResource(
+                  instanceId,
+                  `/api/v1/cgn/status/${fiscalCode}`,
+                  instanceId
+                );
```

Comment:
Could be a problem if this redirect returns an un-updated content? The update happens inside an async operation into the orchestrator (may even fail). Maybe we can return the orchestrator status with [CreateCheckStatusResponse](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.webjobs.extensions.durabletask.idurableorchestrationclient.createcheckstatusresponse?view=azure-dotnet)

Reply:
Hmmm, this redirect points to `getCgnStatus` endpoint, that retrieves data directly from database ( for the CGN status ) so it always returns the last saved status on cosmos, so we can be reasonably sure that it returns always updated content :)